index.js

```diff
@@ -320,23 +320,30 @@ Router.prototype.off = function off (method, path) {
     const pathFull = path.replace(OPTIONAL_PARAM_REGEXP, '$1$2')
     const pathOptional = path.replace(OPTIONAL_PARAM_REGEXP, '$2')
 
-    this.off(method, pathFull)
-    this.off(method, pathOptional)
+    this.off(method, pathFull, derivedConstraints)
+    this.off(method, pathOptional, derivedConstraints)
     return
   }
 
+  function matcher (currentConstraints = {}) {
+    if (!derivedConstraints) return true
+
+    return Object.keys(derivedConstraints)
+      .every((key) => derivedConstraints[key] === currentConstraints[key])
+  }
+
   // Rebuild tree without the specific route
   const ignoreTrailingSlash = this.ignoreTrailingSlash
   var newRoutes = self.routes.filter(function (route) {
     if (!ignoreTrailingSlash) {
-      return !(method === route.method && path === route.path)
+      return !(method === route.method && path === route.path && matcher(route.opts?.constraints))
```

Comment:
The optional chaining operator only appeared in nodejs 14, we can't use it since we support the 12 version.